### PR TITLE
arash/auto-carousel

### PR DESCRIPTION
**Pull request by arash-fs**
# Description

Summary of the change and/or issue that is fixed.

Changes:

- refactor auto carousel with css animation instead of using js
- add ticker as a carousel

## Type of change

-   [x] Bug fix
-   [x] New feature
-   [ ] Breaking change
-   [ ] Dependency update
-   [ ] This change requires a documentation update

### CODE REVIEW

File: src/components/elements/auto-carousel.js [hunk 1]
```diff
@@ -1,92 +1,112 @@
...
     handler = entries => {
         let entry
         for (entry of entries) {
             if (entry.isIntersecting) {
-                this.interval_ref = window.setInterval(
-                    this.handleInterval,
-                    this.props.transition_delay,
-                )
+                this.setState({
+                    shoudl_carousel_move: true,
```
Comment (mustofa-binary) [inline src/components/elements/auto-carousel.js]:
```suggestion
                    should_carousel_move: true,
```

File: src/components/elements/auto-carousel.js [hunk 2]
```diff
@@ -1,92 +1,112 @@
...
+                })
             } else {
-                window.clearTimeout(this.timeout_ref)
-                window.clearInterval(this.interval_ref)
+                this.setState({
+                    shoudl_carousel_move: false,
```
Comment (mustofa-binary) [inline src/components/elements/auto-carousel.js]:
```suggestion
                    should_carousel_move: false,
```

File: src/components/elements/auto-carousel.js [hunk 3]
```diff
@@ -1,92 +1,112 @@
...
-    }
     componentDidMount() {
-        const node = this.my_ref.current
         let observer = new IntersectionObserver(this.handler)
-        observer.observe(node)
+        observer.observe(this.my_ref.current)
+    }
+    static getDerivedStateFromProps(props, state) {
+        if (state.items.length === 0) {
+            const newItems = [...props.children, ...props.children].map(
```
Comment (mustofa-binary) [inline src/components/elements/auto-carousel.js]: shouldn't it be just single `...props.children` ?

File: src/components/elements/auto-carousel.js [hunk 4]
```diff
@@ -1,92 +1,112 @@
...
+            <>
+                <AutoCarouselSection
+                    width={this.props.carousel_width}
+                    ref={this.my_ref}
+                >
+                    {this.state.items !== []
+                        ? this.state.items.map(({ Component, key }) => {
+                              return (
+                                  <ItemContainer
+                                      shoudl_carousel_move={
```
Comment (mustofa-binary) [inline src/components/elements/auto-carousel.js]:
```suggestion
                                      should_carousel_move={
```

File: src/components/elements/auto-carousel.js [hunk 5]
```diff
@@ -1,92 +1,112 @@
...
+                                          this.state.shoudl_carousel_move
```
Comment (mustofa-binary) [inline src/components/elements/auto-carousel.js]:
```suggestion
                                          this.state.should_carousel_move
```

File: src/pages/home/_payment-methods.js [hunk 6]
```diff
@@ -0,0 +1,57 @@
+import React from 'react'
+import styled from 'styled-components'
+import AutoCarousel from 'components/elements/auto-carousel.js'
+import { SectionContainer } from 'components/containers/container'
+import WebMoney from 'images/svg/payment-webmoney.svg'
```
Comment (mustofa-binary) [inline src/pages/home/_payment-methods.js]:
```suggestion
// SVG
import WebMoney from 'images/svg/payment-webmoney.svg'
```

File: src/pages/home/_ticker.js [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+    componentWillUnmount() {
+        BinarySocketBase.send({
+            forget_all: 'ticks',
+        })
+    }
+
+    render() {
+        return (
+            <TickWrapper>
+                <Divider />
```
Comment (mustofa-binary) [inline src/pages/home/_ticker.js]: no need two divider, single divider will do

File: src/pages/home/_ticker.js [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+    }
+}
+
+Tick.propTypes = {
+    display_name: PropTypes.string,
+    symbol: PropTypes.string,
+}
+const Ticker = () => {
+    return (
+        <AutoCarousel carousel_width="100%" transition_duration={20000}>
```
Comment (mustofa-binary) [inline src/pages/home/_ticker.js]: so, every 20 seconds, it will go to the first component? or it will stop?, can you please add a comment here

**Comment by ashkanx** [comment]
@arash-fs please review the comments

File: src/components/elements/auto-carousel.js (hunk 3 above)
Comment (arash-fs) [reply src/components/elements/auto-carousel.js]: No, If we use single `...props.children` when it reaches the end of the `auto-carousel` we will see empty space. so we should render 2 of them right after each other.

File: src/pages/home/_ticker.js (hunk 8 above)
Comment (arash-fs) [reply src/pages/home/_ticker.js]: A whole loop takes 20 seconds,
sure I'll add comments

File: src/components/elements/auto-carousel.js [hunk 9]
```diff
@@ -1,92 +1,115 @@
...
-    max-width: 9.6rem;
-    margin-right: 3.2rem;
+const ItemContainer = styled.div`
+    padding: 0 ${props => props.padding / 2}px;
+    animation-name: ${props =>
+        props.should_carousel_move ? move_items(props.count_child * 100) : ''};
+    animation-duration: ${props => props.transition_duration}ms;
+    animation-fill-mode: both;
+    animation-iteration-count: infinite;
+    animation-timing-function: linear;
```
Comment (oskar-binary) [inline src/components/elements/auto-carousel.js]: Can you try moving this to the parent `AutoCarouselSection`? should fix the "shakiness"

Comment (arash-fs) [reply src/components/elements/auto-carousel.js]: Let me try your suggestion, I'll tell you the result.

File: src/components/elements/dot_loader.js [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+    ::after {
+        color: grey;
+        content: ' .';
+        animation: ${dots} 1s steps(5, end) infinite;
+        font-size: 2rem;
+        line-height: 1;
+    }
+`
+const DotLoader = () => <InitialLoader />
+export default DotLoader
```
Comment (mustofa-binary) [inline src/components/elements/dot_loader.js]: you can export default InitialLoader directly

Comment (arash-fs) [reply src/components/elements/dot_loader.js]: done

**Comment by arash-fs** [comment]
@mustofa-binary This card takes to long. I think it's better to fix this animation in another card

**Pull request merged**